Ambiguous var orpat

maranget opened the pull request:
# A new warning related to pattern matching.

In some occasion the combination of or-patterns and when guards may lead to bugs due
to (wrong) expected semantics.
## Example

Consider the simple type of expressions
`
type exp = Const of int | Var of string | Mult of exp * exp
;;
`
Neutral element is checked by the following function
`
let is_neutral = function
  | 1 -> true
  | _ -> false
;;
`
And, here we go, this fuction is an _intelligent_ constuctor for mult:

```
let mult x y = match x,y with
| (Const i,z)
| (z,Const i)
  when is_neutral i -> z
| _,_ -> Mult (x,y)
;;
```

One may think that `mult (Const 2) (Const 1)` will yield `Const 2`. This is not the case as, according to semantics:
-  The value (Const 2,Const 1) is an instance of the first pattern (Const i,z) and thus the variable `i` is bound to the constant `2`.
- The guard `is_neutral 2` yields false.
- And thus `mult (Const 2) (Const 1)` yields `Mult (Const 2) (Const 1)`.

The new warning detects such a situation:
`
File "t.ml", line 10, characters 2-27:
Warning 57: Ambiguous guarded pattern, variable i may match different or-pattern arguments
`
## Semantics

The new warning lists variables that occur free in a `when` guard and that may be bound to different sub-parts of the matched value by different arguments of or-patterns.
## Significance

The new warning is useful for advanced users and is believed not to occur too frequently. It may draw attention on bugs. It is enabled by default.
## Potential enhancements

Provide a and example of a value that can be matched ambiguously.
Restrict warning to values that are not matched by higher priority patterns as in:

```
let mult x y = match x,y with
| (Const i,Const j) -> Const (i*j)
| (Const i,z)
| (z,Const i)
  when is_neutral i -> z
| _,_ -> Mult (x,y)
```

gasche committed:
minor: add Ident.compare to call Set.Make(Ident) directly

gasche committed:
minor: add Path.heads: find all Ident.t free in a Path.t

Commit:
PR#7031 ambiguous guarded or-pattern.
  - Check proper (in parmatch.ml)
  - Call check just after unused pattern check (in typecore.ml, correct ?)
  - Define and print warning (in warning.ml)

maranget committed:
PR#7031 Fix ambiguous orpats + guards in the compiler.

maranget committed:
PR#7031, Change file + Tests + error message

trefis commented:
Is:

> As a principle of OCaml Pattern Matching : thou shalst not duplicate.

the reason why we are warning instead of having the behavior every one expects?

(This might appear as a salty reaction to your message on my PR, but it is not! It is a genuine question)

maranget commented:
> the reason why we are warning instead of having the behavior every one expects?

The reason is that the semantics of when guards is defined as
- First match pattern,
- Then check guard.

In other word,  guards are not part of patterns they are part of clauses.

Having another definition for guards is not that clear (think about or-patterns that occur deep inside patterns). 

On a related note, and before you tell, having _p_ `when` _e_ to be a pattern is not acceptable either: this would mean huge changes in the compiler for a dubious benefit.

chambart commented:
If I understand correctly what you are saying, the semantics of guards as part of the pattern is not undesirable, but only difficult to implement ?

If this is the case, I imagine that the simpler semantics would have the order, and the number of times a guard is evaluated as undefined with those kind of cases ?

Would you object to someone proposing a patch to rewrite that kind of patterns by duplicating ambiguous match cases before pattern compilation (guard clause would be compiled as a function, and the body as a static raise) ? This has a potential for exponential blowup, but I don't think that this really happen in practice (like exponential cases in the type checker), and we could warn if we start generating too many branches.

alainfrisch commented:
Except for the more difficult compilation, it would indeed be a good idea to have guard nested under or-pattern, I would say.  This would probably lead to relaxing the constraint that the set of bound variables must be identical on both sides of an or-pattern, so as to allow writing:

``` ocaml
  | Some x when x < 0 | None -> ...
```

The identifier `x` would not be bound in the rhs.  More precisely, it should probably be stored in the environment (as a new kind of `Val_*` binding) in order to (i) hide any previously binding on the same name; (ii) enable a reasonable error message if `x` is referenced.

Users would then probably also ask for specifying default values as in:

``` ocaml
  | (Some x when x >= 0) | (None with x = 0) -> ...
```

gasche commented:
@trefis, @chambart : I don't think there is a reasonable way to hide the problem away. The current semantics _is_ defined and reasonable. We cannot silently change it without risking also breaking user expectations -- for example evaluating a guard twice is a no-no in my book. Emitting a warning to discourage this use-case is a much better solution: people can analyse the issue in their code and decide which semantics they want -- by writing the code in a different way, or locally disabling the warning.

@alainfrisch I'm not opposed to the features you propose, but having guards under or-patterns is difficult to implement. The proposed warning can be considered independently; this means that when it fires, duplicating the guard is not an option for users, they have to duplicate the whole clause, but it is not very difficult to do (adding local functions to not duplicate code).

Right now there is code lying around which is probably not doing what its authors expect. Detecting this issue and warning users can help find and fix these cases, and this is what is proposed in the present PR (and the previous version #317 ).

trefis commented:
I didn't expect to spawn such a discussion (but I will happily take part in it!), sorry.

@maranget wrote:

> The reason is that the semantics of when guards is defined as [...]

Fair enough, then I guess my question should have been _"is the reason why we don't change the semantics that... ?"_.

> Having another definition for guards is not that clear (think about or-patterns that occur deep inside patterns).

Since most people seem to have the same intuition of what the behavior should be, I'm sure we could probably come up with a definition.
As for or-patterns occurring deep inside patterns, I don't see how that complicates the matter, could you elaborate?

> On a related note, and before you tell, having p when e to be a pattern is not acceptable either: this would mean huge changes in the compiler for a dubious benefit.

I actually wasn't about to ask for that, but it seems you woke some people up!

@gasche wrote:

> We cannot silently change it without risking also breaking user expectations --for example evaluating a guard twice is a no-no in my book. Emitting a warning to discourage this use-case is a much better solution: people can analyse the issue in their code and decide which semantics they want -- by writing the code in a different way, or locally disabling the warning.

I don't see why we couldn't do both.
Change the semantics and have a warning to inform user that "the behavior has changed since 4.03, here is what happened, here is what happens now".
I will add that at the moment user expectations seems to not be met in any case (this is without a doubt the users fault, but it is true nonetheless). At least, the fact that the warning triggers on the compiler itself does hint so.

The _"evaluating a guard twice is a no-no in my book"_ part seems silly given the context, but that's like my opinion man.

---

That being said, I am glad to see something done about this issue (and so is everyone, I would guess)!
I had a look at the code so I am gladly stepping in @gasche shoes for some minor nitpicks:
- the changes to Changes looks dubious
- `IdSet` (now defined in `parmatch.ml`) already exists in `mtype.ml` and `lambda.ml` (under the name `IdentSet` in these two files). Now might be the time to factorize all these instances?
- (beware: this one is more of a personal style preference than anything) `do_push` and `push_vars` are defined in `parmatch.ml` as mutually recursive while in fact they are not. Do you mind changing that?
- I am sad to see that the content of `filter_all` (`discr_pat`, `filter_rec`, etc) is so slightly different from the other functions of the same already present in the file. (But I don't have anything to propose to factorize the various versions)

As for deeper comments on the code itself (instead of its style) I have none (yet?).

gasche commented:
> As for or-patterns occurring deep inside patterns, I don't see how that complicates the matter, could you elaborate?

In the case of

``` ocaml
| (_, Some x, _) | (_, _, Some x)
  when foo x ->
```

it would seem rather simple to say "we try both branches and test the guard in each case". However, this is much less clear in the case of

``` ocaml
| (_, ((Some x, _) | (_, Some x)))
  when foo x ->
```

because the branches we are speaking of are deep inside the pattern. We are used to think of a pattern `p` as a one-time thing, and have different expectations for `(p | q) when e` because we see a list of alternatives rather a single pattern. But it sounds fishy to start thinking of a pattern `p` with no or-choice at the toplevel as something that can return control several times to a guard.

(The issue would be even worse if we had pattern synonym, such as some or-patterns can be hidden as definitions.)

trefis commented:
> But it sounds fishy to start thinking of a pattern p with no or-choice at the toplevel as something that can return control several times to a guard.

Right, that is the part I disagree on then.
I do get your point though, I was just wondering if there was something I hadn't considered.

gasche commented:
I don't strongly disagree with your proposal to change the semantics and warn about it. The problem is that this sounds very difficult to implement correctly (I don't know whether the pattern-matching compilation machinery can be extended to do it). Unless someone volunteers an implementation, this is idle talking. If we agree that having the warning makes sense even without such a different semantics, then we should be able to discuss the warning in isolation. Then we can wait for the miraculous "guards deep inside patterns" pull request to discuss this part further.

alainfrisch commented:
> @alainfrisch I'm not opposed to the features you propose, but having guards under or-patterns is difficult to implement. The proposed warning can be considered independently

Yes of course.  I think the warning is great and I'm not proposing the change the current semantics.

trefis commented:
> The problem is that this sounds very difficult to implement correctly (I don't know whether the pattern-matching compilation machinery can be extended to do it). Unless someone volunteers an implementation, this is idle talking.

I was just being opportunist, since @maranget is the expert and seems to have a bit of time to dedicate to the issue :)

maranget commented:
> I was just being opportunist, since @maranget is the expert and seems to have a bit of time to dedicate to the issue :)

Ok I see two semantics issues, as far as I understand:
1. Change the semantics of `when` guards so that it matches some intuition in the case of toplevel or-patterns. I have the feeling that nobody has taken the time to define what this semantics would be. I will not do it either.
2.  Add various expressions in patterns, as in-deep guards (ie _p_ `when` _e_ is a pattern)
   or value bindings (_p_ `with` _x_ `=` _e_ is also a pattern). This would somehow enhance pattern matching expressivity, but would also complicate implementation up to unpredictable levels. Consider that
   the type of patterns and of expressions would become mutually recursive, which they are not at the moment. I suspect that such a change would have a major impact on the front end.

maranget commented:
Thank you for your suggestions:

> the changes to Changes looks dubious

Changed.

> IdSet (now defined in parmatch.ml) already exists in mtype.ml and lambda.ml (under the name IdentSet in these two files). Now might be the time to factorize all these instances?

I leave it as it is. No real duplication (one could define a module Set = Set.Make(...) in ident.ml)

> do_push and push_vars are defined in parmatch.ml as mutually recursive while in fact they are not. Do you mind changing that?

Done.

> I am sad to see that the content of filter_all (discr_pat, filter_rec, etc) is so slightly different from the other functions of the same already present in the file. (But I don't have anything to propose to factorize the various versions)

I am as sad as you are.

gasche commented:
I think there _is_ a clear semantics for the multi-return idea: a pattern matches a variable and returns an (ordered) list of environments. The clause is taken if the guarded is true when substituted to one of the environments (tested in order). That is clear, but this clarification does not make it any easier to implement. (Note however that duplicates can only be removed from the environment list, and only cases where these environments matter (not just their existence) is precisely the case of ambiguous guard that we are trying to detect here -- in all other cases the multi-return aspect can be optimized away)

gasche commented on Changes:
```diff
@@ -83,9 +83,13 @@ Compilers:
 - PR#7067: Performance regression in the native compiler for long
   nested structures (Alain Frisch, report by Daniel Bünzli, review
   by Jacques Garrigue)
+- PR#6400: better error message for '_' used as an expression
```
This Changes line results from a merge conflict (it was moved to another place in between and is now duplicated), it should be removed.

maranget replied:
Done, I'd rather no rewrite history for this.  Changes is better changed  at the very end....

trefis replied:
> Changes is better changed at the very end...

But it's not your change to Changes which is at the end, it's a correction to your change to Changes (which was itself done a commit before).
You should squash :)

http://existentialcomics.com/comic/9

maranget replied:
> > - PR#7067: Performance regression in the native compiler for long
> >   nested structures (Alain Frisch, report by Daniel Bünzli, review
> >   by Jacques Garrigue)
> >   +- PR#6400: better error message for '_' used as an expression
> > 
> > Changes is better changed at the very end...
> 
> But it's not your change to Changes which is at the end, it's a correction to your change to Changes (which was itself done a commit before).
> You should squash :)
> Ok squashed.

--Luc

gasche commented on typing/parmatch.ml:
```diff
@@ -2052,3 +2052,186 @@ let check_partial_param do_check_partial do_check_fragile loc casel =
+    (String.concat ","
+       (List.map Ident.name (IdSet.elements vs)))
+
+let pretty_seen rs =
```
These `pretty_` functions are unused, aren't they?

(I think it is better to have the domain logic in parmatch, and the user-printing logic in the file that knows how to print warnings in general.)

maranget replied:
> > +
> > +let pattern_vars p = IdSet.of_list (Typedtree.pat_bound_idents p)
> > +
> > +(\* Row for ambiguous variable search,
> > -   unseen is the traditional pattern row,
> > -   seen   is a list of position bindings *)
> >   +
> >   +type amb_row = { unseen : pattern list ; seen : IdSet.t list; }
> >   +
> >   +
> >   +let pretty_set vs =
> > -  sprintf "{%s}"
> > -    (String.concat ","
> > -       (List.map Ident.name (IdSet.elements vs)))
> >   +
> >   +let pretty_seen rs =
> 
> These `pretty_` functions are unused, aren't they?
> 
> (I think it is better to have the domain logic in parmatch, and the user-printing logic in the file that knows how to print warnings in general.)

Well this is debug leftover code. I erase.

## 

Luc

gasche commented:
The way we compute the free variables of the guard need to be tweaked slightly to take into account ghost bindings generated by the type-checking first-class module patterns (`#modulepat` attribute on `Texp_letmodule` constructs).

maranget committed:
Fix minor bug, avoid douple unpacking in the case of
   (module M:S) | (module M:S) -> ... M ..
That is or pattern whose arguments contain module variables
(AST: Ppat_unpack).

maranget committed:
PR#7031 Handle module variables in patterns.

gasche committed:
comment and refactor the code: now I understand it all!

gasche commented on typing/parmatch.ml:
```diff
@@ -2052,3 +2052,272 @@ let check_partial_param do_check_partial do_check_fragile loc casel =
+    (struct
+      type t = Ident.t
+      let compare = Pervasives.compare
+    end)
```
This could be just `Set.Make(Ident)` now that `Ident` has a `compare` function.

trefis replied:
I will do a PR to remove all the `IdentSet` instances I mentioned earlier once this PR is merged.
So it's probably not worth bothering @maranget about it.

gasche commented:
I reviewed Luc's new implementation and am convinced it is correct. Along the way I added a few comments and clarified some parts.

Unfortunately, we still cannot expect someone not familiar with the usual matrix approach to compile and warn about pattern-matching to understand the relation between the implementation and the specification of the warning. I would like to eventually document that in the compiler, but this is a longer-term project -- and I think this PR could be considered independently.
